File: 0075-sort-colors/0075-sort-colors.js
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var sortColors = function(nums) {
    let current = 0
    let left = 0
    let right = nums.length - 1
    
    while(current <= right) {
        if(nums[current] == 0) {
            [nums[current], nums[left]] = [nums[left], nums[current]]
            current ++
            left ++
        } else if(nums[current] == 2) {
            [nums[current], nums[right]] = [nums[right], nums[current]]
            // current could be 1 or 0
            // if it is a zero, we might want to check the left pointer
            // current ++
            right --
        }
        else
            current ++
    }
    
    return nums
};

File: 0075-sort-colors/0075-sort-colors.java
class Solution {
    public void sortColors(int[] nums) {
        int zeroPtr = 0;
        int twoPtr = nums.length - 1;
        int idx = 0;
        
        while(idx <= twoPtr) {
            if(nums[idx] == 2) {
                nums[idx] = nums[twoPtr];
                nums[twoPtr] = 2;
                twoPtr--;
            } else if(nums[idx] == 0) {
                nums[idx] = nums[zeroPtr];
                nums[zeroPtr] = 0;
                zeroPtr++;
                idx++;
            } else 
                idx++;
            
        }
    }
}